pkg/ebpf/map.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpf
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/lock"
+
+	ciliumebpf "github.com/cilium/ebpf"
+)
+
+type MapSpec = ciliumebpf.MapSpec
+
+const (
+	PerCPUHash = ciliumebpf.PerCPUHash
+
+	PinByName = ciliumebpf.PinByName
+)
+
+// IterateCallback represents the signature of the callback function expected by
+// the IterateWithCallback method, which in turn is used to iterate all the
+// keys/values of a map.
+type IterateCallback func(key, value interface{})
+
+// Map represents an eBPF map.
+type Map struct {
+	*ciliumebpf.Map
+
+	spec       *MapSpec
+	key, value interface{}
+
+	path string
+	lock lock.RWMutex
+}
+
+// NewMap creates a new Map object.
+func NewMap(spec *MapSpec, key, value interface{}) *Map {
```

Comment:
IMO `key, value interface{}` is very unsafe. This allows any value to be written into a BPF map. Think of a potential bug which mixes values written into a map. Debugging such bugs in production could easily turn into a hell.

I'd suggest to introduce an interface type for key and value (same as we have in `pkg/bpf` today).

Reply:
These `key` and `value` interfaces are only used to derive a type from during `IterateWithCallback()`. The values specified here are never written to the map.

@brb your comment aligns with my suggestion of avoiding a callback-based model. IMO the iteration API should be limited to `Map.Next()` and then reading keys and values using `Map.{Key,Value}Uint16()`, `Map.{Key,Value}Bytes()` etc.

Reply:
> These key and value interfaces are only used to derive a type from during IterateWithCallback(). The values specified here are never written to the map.

Yeah, but the returned map allows `(*cilium/ebpf.Map).Update(key, value interface{}, flags MapUpdateFlags) error` to be invoked which accepts any key/val.

> IMO the iteration API should be limited to Map.Next() and then reading keys and values using Map.{Key,Value}Uint16(), Map.{Key,Value}Bytes() etc.

I think it might be tricky to implement this correctly. See https://github.com/cilium/cilium/blob/master/pkg/bpf/map_linux.go#L665 for all hacks.

Reply:
> IMO key, value interface{} is very unsafe. This allows any value to be written into a BPF map. Think of a potential bug which mixes values written into a map. Debugging such bugs in production could easily turn into a hell.

That's a fair concern :+1: on the other hand I think the user of a map package like `maps/metricsmap` or `maps/lbmap` should not be concerned about this: the map package should handle (and hide) these details and expose only safe methods (ideally the user of `maps/*` should know nothing about the `ebpf` package). So once it's done and tested in the map package, there shouldn't be any risk of mixing different types

> I'd suggest to introduce an interface type for key and value (same as we have in pkg/bpf today).

I was trying to avoid that, if possible, as I think it introduces a good amount of complexity for little benefit (but I can only talk for the metricsmap case so far, I'm not sure if this will still hold when we start exposing the `Iterate()`/`Next()` methods)

> These key and value interfaces are only used to derive a type from during IterateWithCallback(). 

Good point. I think I'll move the logic to pass these key/value interfaces around directly in the `IterateWithCallback()` implementation of a given map package, to simplify the signature of `ebpf.NewMap()` (and also to save us a few calls to the `reflect` package)

Reply:
Just reading through it again, could we maybe do away with `IterateWithCallback()`? We should probably add something like `(*MapIterator).Dump()` and `(*MapIterator).AtomicDump()` (based on https://github.com/cilium/cilium/blob/master/pkg/bpf/map_linux.go#L665) to `cilium/ebpf` and create (or generate) wrappers that add typing in our internal subpackages (metricsmap, lbmap).

Just not a fan of the callback model in Go. :(

Reply:
I'm not sure I fully understand what you are proposing with `(*MapIterator).Dump()` and `(*MapIterator).AtomicDump()`, could you elaborate a bit more on that?

> Just not a fan of the callback model in Go. :(

I fully share the sentiment, but at the same time I don't think this should be a blocker until we hit a case that requires more control or atomicity over the iteration.

I think the current use case for metrics fits nicely with the `IterateWithCallback()` approach as (in my opinion):
* it doesn't make the code hard to follow:
https://github.com/cilium/cilium/blob/a096e381708d0585cb02aa331c06b06a05059d97/cilium/cmd/bpf_metrics_list.go#L72-L76 https://github.com/cilium/cilium/blob/a096e381708d0585cb02aa331c06b06a05059d97/pkg/maps/metricsmap/metricsmap.go#L187-L189
* it doesn't require any special handling (after all, we are just dumping metrics from a map which does never get an element removed)
* and also it's not sitting in any hot path

Reply:
> I think the current use case for metrics fits nicely with the `IterateWithCallback()` approach

I agree, let's port first and upstream changes later as we see common patterns emerge that would fit the external lib.

> I'm not sure I fully understand what you are proposing with `(*MapIterator).Dump()` and `(*MapIterator).AtomicDump()`

Oh, it just looks to me like some of this code could be simplified if `cilium/ebpf` map objects would have a method that returns their contents as a `map[interface{}]interface{}` (or `[]byte`s, ..) instead of the caller having to deal with iterators and having to write callbacks, etc..

Reply:
> That's a fair concern +1 on the other hand I think the user of a map package like maps/metricsmap or maps/lbmap should not be concerned about this: the map package should handle (and hide) these details and expose only safe methods (ideally the user of maps/* should know nothing about the ebpf package). So once it's done and tested in the map package, there shouldn't be any risk of mixing different types

I think that a bug in a higher-level maps package is still possible. Also, out of laziness for a simple map it will be tempting not to introduce the higher-level package.

Reply:
Let me take a step back as I think this might have gone missed in the thread: I removed the `key, value interface{}` parameters from the `NewMap` method as we only need them in `IterateWithCallback()` (so I made that function responsible for handling the type conversions, which also saves us also a couple of call to the `reflect` package).

> I think that a bug in a higher-level maps package is still possible. 

Consider for example the `IterateWithCallback` method for `metricsmap`:

https://github.com/cilium/cilium/blob/7b780cc746785e09763723b2002e69113c1415d7/pkg/maps/metricsmap/metricsmap.go#L99-L106

whrere `IterateCallback` is defined as
https://github.com/cilium/cilium/blob/7b780cc746785e09763723b2002e69113c1415d7/pkg/maps/metricsmap/metricsmap.go#L33

I'm failing to see a way to have a mismatch for the `Key`/`Value` types as the callback is explicitly typed to receive a `*metricsmap.Key` and a `*metricsmap.Value` (and also how having a `bpf.Key` and `bpf.Value` interface would help catch these types of errors).

I think the same would hold for other methods that work with keys and values, for example I can imagine how the `Update()` method for the maglev map would look something like:

```go
func (m *MaglevInnerMap) Update(key *MaglevInnerKey, val *MaglevInnerVal) {
    innerMap.Update(innerKey, innerVal, 0);
}
```

so again, a small layer of abstraction could prevent types mismatch.

> Also, out of laziness for a simple map it will be tempting not to introduce the higher-level package.

I really don't think we should design this taking laziness into account. Having a per-map higher level package gives us:

* type safety as the user is forced to use the correct keys and values types (as opposed to using directly `cilium/ebpf`, where all functions which work with maps expect `interface{}`s)
* less code duplication
* a consistent interface across the different map packages

Reply:
Thanks for the explanation. Any idea how we can prevent "direct" usage of underlying `cilium/ebpf` maps without the small layer of abstraction?

Reply:
> Any idea how we can prevent "direct" usage of underlying cilium/ebpf maps without the small layer of abstraction?

I don't have any suggestion to actually enforce that beside adding some documentation on the package and code review :disappointed:

Reply:
`[ "$(grep github.com/cilium/ebpf pkg/ | grep -v pkg/ebpf | wc -l)" != "0" ]` ?

Reply:
Nice! That would work :boom: (although we couldn't enable this right now as there are a few packages which are already using `cilium/ebpf` and need to be updated).

Should this check live in `contrib/scripts`, so be part of the Go prechecks?